src/github.com/matrix-org/dendrite/cmd/syncserver-integration-tests/main.go

```diff
@@ -0,0 +1,526 @@
+	cmd := exec.Command(
+		filepath.Join(kafkaDir, "bin", "kafka-console-producer.sh"),
+		"--broker-list", kafkaURI,
+		"--topic", topic,
+	)
+	// Send stdout and stderr to our stderr so that we see error messages from
+	// the kafka process.
+	cmd.Stdout = os.Stderr
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = strings.NewReader(strings.Join(data, "\n"))
+	return cmd.Run()
+}
+
+// TODO: dupes roomserver integration tests. Factor out.
+func deleteTopic(topic string) error {
+	cmd := exec.Command(
+		filepath.Join(kafkaDir, "bin", "kafka-topics.sh"),
+		"--delete",
+		"--if-exists",
+		"--zookeeper", zookeeperURI,
+		"--topic", topic,
+	)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stderr
+	return cmd.Run()
+}
+
+// TODO: dupes roomserver integration tests. Factor out.
+func canonicalJSONInput(jsonData []string) []string {
+	for i := range jsonData {
+		jsonBytes, err := gomatrixserverlib.CanonicalJSON([]byte(jsonData[i]))
+		if err != nil {
+			panic(err)
+		}
+		jsonData[i] = string(jsonBytes)
+	}
+	return jsonData
+}
+
+func doSyncRequest(done chan error, want []string, since string) func() {
+	return func() {
+		cli := &http.Client{
+			Timeout: 5 * time.Second,
+		}
+		res, err := cli.Get("http://" + syncserverAddr + "/api/_matrix/client/r0/sync?access_token=@alice:localhost&since=" + since)
+		if err != nil {
+			done <- err
+			return
+		}
+		if res.StatusCode != 200 {
+			done <- fmt.Errorf("/sync returned HTTP status %d", res.StatusCode)
+			return
+		}
+		defer res.Body.Close()
+		resBytes, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			done <- err
+			return
+		}
+		jsonBytes, err := gomatrixserverlib.CanonicalJSON(resBytes)
+		if err != nil {
+			done <- err
+			return
+		}
+		if string(jsonBytes) != want[0] {
+			fmt.Println("Expected:")
+			fmt.Println(want[0])
+			fmt.Println("Got:")
+			fmt.Println(string(jsonBytes))
+			done <- fmt.Errorf("/sync returned wrong bytes")
+			return
+		}
+		// all good, clean up
+		close(done)
+	}
+}
+
+func testSyncServer(input, want []string, since string) {
+	deleteTopic(inputTopic)
+	if err := createTopic(inputTopic); err != nil {
+		panic(err)
+	}
+	if err := writeToTopic(inputTopic, canonicalJSONInput(input)); err != nil {
+		panic(err)
+	}
+
+	if err := createDatabase(testDatabaseName); err != nil {
+		panic(err)
+	}
+
+	const configFileName = "sync-api-server-config-test.yaml"
+	err := ioutil.WriteFile(configFileName, []byte(syncServerConfigFileContents), 0644)
+	if err != nil {
+		panic(err)
+	}
+
+	cmd := exec.Command(
+		filepath.Join(filepath.Dir(os.Args[0]), "dendrite-sync-api-server"),
+		"--config", configFileName,
+		"--listen", syncserverAddr,
+	)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stderr
+	if err := cmd.Start(); err != nil {
+		panic("failed to start sync server: " + err.Error())
+	}
+	done := make(chan error, 1)
+
+	// TODO: Waiting 1s is racey. Maybe keep hitting it until it doesn't get Connection Refused?
+	time.AfterFunc(1*time.Second, doSyncRequest(done, want, since))
```

Comment:
It should wait until the messages you inserted with kafka come down /sync.
Otherwise your HTTP request will race with the sync server processing the kafka stream.